tool/canonical/scorecard.dart

```diff
@@ -0,0 +1,403 @@
+    details.forEach((detail) => sb.write(detail.header.markdown));
+    sb.write(' |\n');
+
+    if (sorter != null) {
+      scores.sort(sorter);
+    }
+
+    // Body.
+    forEach((lint) => sb.write('${lint.toMarkdown(details)}\n'));
+    return sb.toString();
+  }
+
+  void forEach(void Function(LintScore element) f) {
+    scores.forEach(f);
+  }
+
+  void removeWhere(bool Function(LintScore element) test) {
+    scores.removeWhere(test);
+  }
+
+  static Future<ScoreCard> calculate() async {
+    var lintsWithFixes = await _getLintsWithFixes();
+    var lintsWithBulkFixes = await _getLintsWithBulkFixes();
+    var lintsWithAssists = await _getLintsWithAssists();
+
+    // var issues = await _getIssues();
+    // var bugs = issues.where(_isBug).toList();
+    var bugs = <Issue>[];
+
+    var scoreRuleset = _readScoreLints();
+    var recommendRuleset = _readRecommendLints();
+
+    var scorecard = ScoreCard();
+    for (var lint in registeredLints) {
+      var ruleSets = <String>[];
+      if (scoreRuleset.contains(lint.name)) {
+        ruleSets.add('score');
+      }
+      if (recommendRuleset.contains(lint.name)) {
+        ruleSets.add('recommend');
+      }
+
+      if (ruleSets.isEmpty) {
+        continue;
+      }
+
+      var bugReferences = <String>[];
+      for (var bug in bugs) {
+        if (bug.title.contains(lint.name)) {
+          bugReferences.add('#${bug.number.toString()}');
+        }
+      }
+
+      var lintName = lint.name;
+      scorecard.add(LintScore(
+        name: lintName,
+        hasFix: lintsWithFixes.contains(lintName) ||
+            lintsWithAssists.contains(lintName),
+        hasBulkFix: lintsWithBulkFixes.contains(lintName),
+        maturity: lint.maturity.name,
+        ruleSets: ruleSets,
+        bugReferences: bugReferences,
+      ));
+    }
+
+    return scorecard;
+  }
+
+  // static Future<List<Issue>> _getIssues() async {
+  //   var github = GitHub();
+  //   var slug = RepositorySlug('dart-lang', 'linter');
+  //   try {
+  //     return github.issues.listByRepo(slug).toList();
+  //   } on Exception catch (e) {
+  //     print('exception caught fetching github issues');
+  //     print(e);
+  //     print('(defaulting to an empty list)');
+  //     return Future.value(<Issue>[]);
+  //   }
+  // }
+
+  static Future<List<String>> _getLintsWithAssists() async {
+    var client = http.Client();
+    var req = await client.get(
+        'https://raw.githubusercontent.com/dart-lang/sdk/master/pkg/analysis_server/lib/src/services/correction/assist.dart');
+    var parser = CompilationUnitParser();
+    var cu = parser.parse(contents: req.body, name: 'assist.dart');
+    var assistKindClass = cu.declarations.firstWhere(
+        (m) => m is ClassDeclaration && m.name.name == 'DartAssistKind');
+
+    var collector = _AssistCollector();
+    assistKindClass.accept(collector);
+    return collector.lintNames;
+  }
+
+  static Future<List<String>> _getLintsWithBulkFixes() async {
```

Comment:
I don't understand why we're parsing the file. Why not just get the lint names by writing `BulkFixProcessor.lintProducerMap.keys.toList()`? (Same below for `_getLintsWithFixes` using the map in `FixProcessor`.)

Reply:
See above.